File: nanopdf-js/examples/03-rendering.ts
/**
 * Example 3: Page Rendering
 *
 * This example demonstrates how to:
 * - Render pages to images
 * - Control resolution and quality
 * - Save as PNG
 * - Create thumbnails
 */

import { Document, Matrix, Colorspace } from 'nanopdf';
import { resolve } from 'path';
import { writeFileSync } from 'fs';

function renderSinglePage() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/hello-world.pdf');
  const doc = Document.open(pdfPath);

  console.log('=== Rendering Single Page ===\n');

  const page = doc.loadPage(0);

  // Render at different resolutions
  const resolutions = [
    { name: 'Thumbnail', dpi: 36, scale: 0.5 },
    { name: 'Screen', dpi: 96, scale: 1.0 },
    { name: 'Print', dpi: 300, scale: 4.0 }
  ];

  for (const res of resolutions) {
    console.log(`Rendering at ${res.name} quality (${res.dpi} DPI)...`);

    const matrix = Matrix.scale(res.scale, res.scale);
    const pixmap = page.toPixmap(matrix, Colorspace.deviceRGB(), false);

    console.log(`  Size: ${pixmap.width} × ${pixmap.height} pixels`);
    console.log(`  Components: ${pixmap.components}`);
    console.log(`  Memory: ${(pixmap.width * pixmap.height * pixmap.components / 1024 / 1024).toFixed(2)} MB`);

    const pngData = page.toPNG(res.dpi);
    const outputPath = resolve(__dirname, `output-${res.name.toLowerCase()}.png`);
    writeFileSync(outputPath, pngData);

    console.log(`  ✅ Saved to: ${outputPath}\n`);
  }

  page.drop();
  doc.close();
}

function createThumbnails() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/multi-page.pdf');
  const doc = Document.open(pdfPath);

  console.log('=== Creating Thumbnails ===\n');

  const thumbnailScale = 0.2; // 20% of original size
  const matrix = Matrix.scale(thumbnailScale, thumbnailScale);

  for (let i = 0; i < Math.min(doc.pageCount, 5); i++) {
    const page = doc.loadPage(i);
    const pixmap = page.toPixmap(matrix, Colorspace.deviceRGB(), false);

    console.log(`Page ${i + 1}: ${pixmap.width}×${pixmap.height} pixels`);

    const pngData = page.toPNG(36); // 36 DPI for thumbnail
    const outputPath = resolve(__dirname, `thumbnail-page-${i + 1}.png`);
    writeFileSync(outputPath, pngData);

    console.log(`  ✅ Saved: ${outputPath}`);

    page.drop();
  }

  doc.close();
  console.log('\n✅ Done!');
}

function renderWithColorspace() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/hello-world.pdf');
  const doc = Document.open(pdfPath);

  console.log('\n=== Rendering with Different Colorspaces ===\n');

  const page = doc.loadPage(0);
  const matrix = Matrix.scale(1.0, 1.0);

  const colorspaces = [
    { name: 'RGB', cs: Colorspace.deviceRGB() },
    { name: 'Gray', cs: Colorspace.deviceGray() },
    { name: 'CMYK', cs: Colorspace.deviceCMYK() }
  ];

  for (const { name, cs } of colorspaces) {
    console.log(`Rendering in ${name} colorspace...`);

    const pixmap = page.toPixmap(matrix, cs, false);

    console.log(`  Components: ${pixmap.components}`);
    console.log(`  Size: ${pixmap.width}×${pixmap.height}`);

    const outputPath = resolve(__dirname, `output-${name.toLowerCase()}.png`);
    const pngData = page.toPNG(96);
    writeFileSync(outputPath, pngData);

    console.log(`  ✅ Saved: ${outputPath}\n`);
  }

  page.drop();
  doc.close();
}

function renderWithAlpha() {
  const pdfPath = resolve(__dirname, '../../test-pdfs/simple/hello-world.pdf');
  const doc = Document.open(pdfPath);

  console.log('\n=== Rendering with Alpha Channel ===\n');

  const page = doc.loadPage(0);
  const matrix = Matrix.scale(2.0, 2.0);

  // Render with alpha channel (transparency)
  const pixmap = page.toPixmap(matrix, Colorspace.deviceRGB(), true);

  console.log(`Rendered with alpha: ${pixmap.width}×${pixmap.height}`);
  console.log(`Components: ${pixmap.components} (includes alpha)`);
  console.log(`Has alpha: ${pixmap.hasAlpha}`);

  const pngData = page.toPNG(144);
  const outputPath = resolve(__dirname, 'output-with-alpha.png');
  writeFileSync(outputPath, pngData);

  console.log(`✅ Saved: ${outputPath}`);

  page.drop();
  doc.close();
}

// Run examples
if (require.main === module) {
  try {
    renderSinglePage();
    createThumbnails();
    renderWithColorspace();
    renderWithAlpha();
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

export { renderSinglePage, createThumbnails, renderWithColorspace, renderWithAlpha };

File: go-micropdf/examples/03_rendering.go
// Example 3: Page Rendering
//
// This example demonstrates how to:
//   - Render pages to images
//   - Control resolution and quality
//   - Save as PNG
//   - Create thumbnails
//   - Use custom transformations

//go:build example

package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	micropdf "github.com/lexmata/micropdf/go-micropdf"
)

func main() {
	ctx := micropdf.NewContext()
	if ctx == nil {
		log.Fatal("Failed to create context")
	}
	defer ctx.Drop()

	pdfPath := filepath.Join("..", "..", "test-pdfs", "simple", "hello-world.pdf")
	doc, err := micropdf.OpenDocument(ctx, pdfPath)
	if err != nil {
		log.Fatalf("Failed to open PDF: %v", err)
	}
	defer doc.Drop()

	page, err := doc.LoadPage(0)
	if err != nil {
		log.Fatalf("Failed to load page: %v", err)
	}
	defer page.Drop()

	// Render at different resolutions
	fmt.Println("=== Rendering at Different Resolutions ===\n")

	type Resolution struct {
		Name string
		DPI  int32
	}

	resolutions := []Resolution{
		{"Thumbnail", 72},
		{"Screen", 150},
		{"Print", 300},
	}

	for _, res := range resolutions {
		fmt.Printf("Rendering at %s quality (%d DPI)...\n", res.Name, res.DPI)

		pngData, err := page.RenderToPNG(res.DPI)
		if err != nil {
			log.Printf("Error rendering: %v", err)
			continue
		}

		outputFile := fmt.Sprintf("output-%s.png", res.Name)
		err = os.WriteFile(outputFile, pngData, 0644)
		if err != nil {
			log.Printf("Error saving: %v", err)
			continue
		}

		fmt.Printf("  ✅ Saved to: %s (%.2f KB)\n\n",
			outputFile, float64(len(pngData))/1024)
	}

	// Custom rendering with pixmap
	fmt.Println("=== Custom Rendering with Transform ===\n")

	// Create 2x scale transform
	scale := micropdf.MatrixScale(2.0, 2.0)

	pix, err := page.RenderToPixmap(scale, false)
	if err != nil {
		log.Fatalf("Failed to render to pixmap: %v", err)
	}
	defer pix.Drop()

	width, _ := pix.Width()
	height, _ := pix.Height()
	n, _ := pix.N()

	fmt.Printf("Rendered pixmap: %dx%d pixels\n", width, height)
	fmt.Printf("Components per pixel: %d\n", n)

	samples, err := pix.Samples()
	if err != nil {
		log.Printf("Error getting samples: %v", err)
	} else {
		fmt.Printf("Pixel data: %d bytes (%.2f MB)\n",
			len(samples), float64(len(samples))/1024/1024)
	}

	// Save pixmap as PNG
	pngData, err := pix.WritePNG()
	if err != nil {
		log.Printf("Error converting to PNG: %v", err)
	} else {
		outputFile := "output-2x.png"
		err = os.WriteFile(outputFile, pngData, 0644)
		if err != nil {
			log.Printf("Error saving: %v", err)
		} else {
			fmt.Printf("✅ Saved to: %s\n", outputFile)
		}
	}

	fmt.Println("\n✅ Done!")
}
